### DISCUSSION TITLE
Class selectors or query language to find elements

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1923] [Ideas / Feature Requests]
Following up several earlier discussions (#1913, #1583), I made a proof of concept class selector for nicegui.  It retrieves all nodes on the page whose html classes (not python classes) match the selector text.

You can chain any number of classnames in the selector text separated by dot.  It's like a rudimentary CSS selector, where dot means descended from (direct parent not required).  Example:
```py
elements = findall ('foo.bar')    # find all elements with class bar under an element of class foo
```

[Demo file is here](https://github.com/ed2050/nicegui-docs/blob/main/demoid.py) if you want to run it.  Relevant code excerpted at end of this post.

I know very little about nicegui internals.  it's likely my approach overlooked some things, got some things wrong, or causes problems in other ways.  Any feedback is very welcome.

# Purpose
To show how nicegui elements can be managed and retrieved across an application.  Ideally nicegui could provide a similar type of functionality in the future for selecting nodes from the element tree.

# Requirements
I and others want a way to retrieve aribtrary nodes from the ui element tree.  It's hard to manage large apps by assigning variable names to each ui component, so you can access them when needed.  I want:
- user-created identifiers or labels for different components
- doesn't need to be globally unique, may have duplicates (eg several elements in different parts of the app can have the label 'searchbutton')
- follows the tree structure of nicegui elements so identifiers can be chained, like foo.bar to find bar elements that are under a foo element.

For this demo, I settled on using id strings set with element.classes().  Because:
- nicegui already stores _classes with each element, no need to modify or subclass nicegui types 
- can set multiple values with .classes, unlike id
- .classes() accepts arbitrary identifier strings
- as long as we avoid class names already used by tailwind / quasar / custom css, extra classnames won't have any effect on rendered page

# Limitations
There are many.  This is just a proof of concept.
- requires calling init func to set root node.  not sure where nicegui stores the root.
- not sure if root node (elt.parent_slot.parent... until parent_slot is none) is per page or covers the entire app.
- selector format is very basic.  could be improved to be more like css selectors (or even same format).
- using .classes() as selector ids requires care to avoid clashing with existing class names for tailwind, quasar.
- could add syntax to select on more fields, such as .props() values, nicegui python class, etc.  imagine `findall ('ui_card:foo@wfull')` to find all elements of type ui.card with class foo and property wfull.
- selection algorithm isn't very efficient.  for each component in selector, algorithm walks the full subtree looking for matches.  worst case performance is on the order of `O(n^x)` where n is nodes in the tree and x is number of selector components.  could be improved to make one pass through the tree instead.  fine for most cases though, where n and x aren't particularly large.
- uses some internals of nicegui like _classes and _collect_descendants () that may change behavior in future

# Takeaways
I hope this is useful to demonstrate what I and others want.  If this concept can be improved and incorporated in nicegui, that would be fantastic.  If not, perhaps others of us can develop this further as an extra layer on top of nicegui (but native support would work better).

# Code Excerpt
[Complete file is here](https://github.com/ed2050/nicegui-docs/blob/main/demoid.py)
```py
from nicegui import ui
from nicegui import native_mode

true, false, none = True, False, None
DEBUG = false

# ------------------------------------------------

class container (object) : pass

g = container ()
g.root = none     # page root to start searches from

# ------------------------------------------------

def set_root (node) :
	'initialize root node for all ui elements'
	
	while node.parent_slot :
		node = node.parent_slot.parent

	g.root = node
	return node

# ------------------------------------------------

def findall (searchids, startnode = none) :
	'return a list of all nodes under startnode (default: root) matching dot-separated searchids'

	if startnode is none : 
		startnode = g.root

	classlist = searchids.split ('.')

	# --- gather all nodes under startnode
	# use a set so each node only appears once; assumes ui elements implement ==

	checknodes = set (startnode._collect_descendants ())
	checknodes.add (startnode)

	matches = []

	while classlist :

		# --- match next item in classlist 

		klass = classlist.pop (0)
		matches = [ x for x in checknodes if klass in x._classes ]

		DEBUG and print (f'{ klass = } : found { len (matches) } : { matches }')

		# --- are we done ?

		if not len (classlist)  : break
			
		# --- update for next item in classlist
		# not efficient to rewalk subtree on each iteration, improve later if needed

		checknodes = set ()
		for node in matches :
			checknodes.add (node)
			checknodes.update (node._collect_descendants ())
			desc = node._collect_descendants ()

			DEBUG and print (f'found { len (desc) } descendants : { desc }')

		matches = []

	return matches

# ------------------------------------------------

def mainpage () : 

   # --- sample elements : two cards each with two buttons

    with ui.card ().classes ('card1') :
        ui.button ('Foo 1').classes ('foo btn')
        ui.button ('Bar 1').classes ('bar btn')

    with ui.card ().classes ('card2') :
        ui.button ('Foo 2').classes ('foo btn')
        ui.button ('Bar 2').classes ('bar btn')

    setroot (ui.label (''))    # dummy element to initialize root

    # --- examples of findall
    buttons = findall ('btn')    # get all buttons
    buttons = findall ('foo')    # get all foo buttons
    buttons = findall ('bar')    # get all bar buttons
    buttons = findall ('card1.foo')    # get all foo buttons on card1
   
```

**Comment by rodja** [comment]
Thank you for this creative and good expressed idea. I think this direction is much better than allowing to set `id` (discussed in #1583). To avoid name clashes with Tailwind/Quasar we could introduce a new `name` parameter (`ui.label("Some Title", name="title")`) or a `.tags` property (`ui.label("Some Title").tags("title header")`) if we want to allow multiple identifiers. Either approach has also the advantage of allowing us to define our own (hopefully more Pythonic) search language and indexing engine. Maybe something similar to tortoise-orm's `filter` or BeautifulSoup's `find_all`:

```py
buttons = ui.get(type=ui.button)
headers = ui.get(tag='header')  
action_buttons = ui.get(type=ui.button, tag='action')
buttons_with_text = ui.get(type=ui.button, text='Click me')
buttons_witout_text = ui.get(type=ui.button).without(text='Click me')
buttons_within_header = ui.get(type=ui.button).within(tag='header')
buttons_within_header_or_footer = ui.get(type=ui.button).within(tag=['header', 'footer'])
```

The object which is returned by `ui.get` should be something like a pseudo-element which allows setting props,classes,style on all containing elements at once. And it should allow of course iterating over all elements:

```py
for element in ui.get(tag='important'):
    element.classes('text-bold')
```

If `type=` is used the iterator automatically casts to this type:

```py
for button in ui.get(tag='important', type=ui.button):
    button.icon = 'priority_high'
```

**Reply by ed2050** [reply]
Thansk rodja.  I like the direction of most of your suggested improvements.

Just one issue: why return a pseudo-element instead of the real element?  I want to do more than just set styles / classes / props.  I may want to:
- change the handler on certain elements.  
- move the element to another place in the tree.  
- check the elements' values (eg see which input fields are empty), without knowing which variable this particular input is bound to.  

And more.  To really be useful, I need access to the full elements.

**Reply by rodja** [reply]
It must be a pseudo-element (or say "group") because the `ui.get()` is a group of elements which needs to provide an iterator interface and the builder-pattern methods `without`, `with`, `within`, ...
You can access each element by using an iterator.

**Reply by rodja** [reply]
I've started a rough implementation in https://github.com/zauberzeug/nicegui/pull/1931. Please have a look and tell me what you are thinking.
